Mathlib/Data/TypeVec.lean

```diff
@@ -0,0 +1,879 @@
+
+theorem splitFun_comp {α₀ α₁ α₂ : TypeVec (n + 1)} (f₀ : drop α₀ ⟹ drop α₁)
+    (f₁ : drop α₁ ⟹ drop α₂) (g₀ : last α₀ → last α₁) (g₁ : last α₁ → last α₂) :
+    splitFun (f₁ ⊚ f₀) (g₁ ∘ g₀) = splitFun f₁ g₁ ⊚ splitFun f₀ g₀ :=
+  eq_of_drop_last_eq rfl rfl
+#align typevec.split_fun_comp TypeVec.splitFun_comp
+
+theorem appendFun_comp_splitFun {α γ : TypeVec n} {β δ : Type _} {ε : TypeVec (n + 1)}
+          (f₀ : drop ε ⟹ α)
+          (f₁ : α ⟹ γ)
+          (g₀ : last ε → β)
+          (g₁ : β → δ) :
+   appendFun f₁ g₁ ⊚ splitFun f₀ g₀
+      = splitFun (α':=γ.append1 δ) (f₁ ⊚ f₀) (g₁ ∘ g₀)
+  :=
+  (splitFun_comp _ _ _ _).symm
+#align typevec.append_fun_comp_split_fun TypeVec.appendFun_comp_splitFun
+
+theorem appendFun_comp  {α₀ α₁ α₂ : TypeVec n}
+                        {β₀ β₁ β₂ : Type _}
+                        (f₀ : α₀ ⟹ α₁) (f₁ : α₁ ⟹ α₂)
+                        (g₀ : β₀ → β₁) (g₁ : β₁ → β₂) :
+    (f₁ ⊚ f₀ ::: g₁ ∘ g₀) = (f₁ ::: g₁) ⊚ (f₀ ::: g₀) :=
+  eq_of_drop_last_eq rfl rfl
+#align typevec.append_fun_comp TypeVec.appendFun_comp
+
+theorem appendFun_comp' {α₀ α₁ α₂ : TypeVec n} {β₀ β₁ β₂ : Type _}
+                        (f₀ : α₀ ⟹ α₁) (f₁ : α₁ ⟹ α₂)
+                        (g₀ : β₀ → β₁) (g₁ : β₁ → β₂) :
+    (f₁ ::: g₁) ⊚ (f₀ ::: g₀) = (f₁ ⊚ f₀ ::: g₁ ∘ g₀) :=
+  eq_of_drop_last_eq rfl rfl
+#align typevec.append_fun_comp' TypeVec.appendFun_comp'
+
+theorem nilFun_comp {α₀ : TypeVec 0} (f₀ : α₀ ⟹ Fin2.elim0) : nilFun ⊚ f₀ = f₀ :=
+  funext fun x => by apply Fin2.elim0 x -- porting note: `by apply` is necessary?
+#align typevec.nilFun_comp TypeVec.nilFun_comp
+
+theorem appendFun_comp_id {α : TypeVec n} {β₀ β₁ β₂ : Type _} (g₀ : β₀ → β₁) (g₁ : β₁ → β₂) :
+    (@id _ α ::: g₁ ∘ g₀) = (id ::: g₁) ⊚ (id ::: g₀) :=
+  eq_of_drop_last_eq rfl rfl
+#align typevec.append_fun_comp_id TypeVec.appendFun_comp_id
+
+@[simp]
+theorem dropFun_comp {α₀ α₁ α₂ : TypeVec (n + 1)} (f₀ : α₀ ⟹ α₁) (f₁ : α₁ ⟹ α₂) :
+    dropFun (f₁ ⊚ f₀) = dropFun f₁ ⊚ dropFun f₀ :=
+  rfl
+#align typevec.drop_fun_comp TypeVec.dropFun_comp
+
+@[simp]
+theorem lastFun_comp {α₀ α₁ α₂ : TypeVec (n + 1)} (f₀ : α₀ ⟹ α₁) (f₁ : α₁ ⟹ α₂) :
+    lastFun (f₁ ⊚ f₀) = lastFun f₁ ∘ lastFun f₀ :=
+  rfl
+#align typevec.last_fun_comp TypeVec.lastFun_comp
+
+theorem appendFun_aux {α α' : TypeVec n} {β β' : Type _} (f : (α ::: β) ⟹ (α' ::: β')) :
+    (dropFun f ::: lastFun f) = f :=
+  eq_of_drop_last_eq rfl rfl
+#align typevec.append_fun_aux TypeVec.appendFun_aux
+
+theorem appendFun_id_id {α : TypeVec n} {β : Type _} :
+    (@TypeVec.id n α ::: @_root_.id β) = TypeVec.id :=
+  eq_of_drop_last_eq rfl rfl
+#align typevec.append_fun_id_id TypeVec.appendFun_id_id
+
+instance subsingleton0 : Subsingleton (TypeVec 0) :=
+  ⟨fun a b => funext fun a => by apply Fin2.elim0 a⟩ -- porting note: `by apply` necessary?
+#align typevec.subsingleton0 TypeVec.subsingleton0
+
+/- porting note: just omitting this `simp` attribute declaration for now
```

Comment:
Why?

If we do this, there should be porting notes on the `-- @[typevec]` below, so we can find them again later.

Reply:
I managed to get the simp attribute working, but had to put it in a separate file (Data.TypeVec.Attr); if I kept the declaration in the original file, it complained that `typevec` is not an attribute on the lower lines.

It seems a bit overkill to have a file for just defining this single attribute, are there any conventions on where to put it otherwise?

Reply:
I think putting it in a new file is fine.